File: 349 Intersection of Two Arrays/349.py
class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        return list(set(nums1) & set(nums2))
    
    def intersection2(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()
        res = []
        l, r = 0, 0
        while l < len(nums1) and r < len(nums2):
            a, b = nums1[l], nums2[r]
            if a < b:
                l += 1
            elif a > b:
                r += 1
            else:
                res.append(a)
                l += 1
                r += 1
                # if there might be duplicated
                while l < len(nums1) and a == nums1[l]:
                    l += 1
                while r < len(nums2) and b == nums2[r]:
                    r += 1
        return res

File: 349 Intersection of Two Arrays/349.java
class Solution {
    // Solution 1: Hash Set
    public int[] intersection(int[] nums1, int[] nums2) {
        List<Integer> a = toList(nums1);
        List<Integer> b = toList(nums2);
        Set<Integer> set = new HashSet<>(a);
        set.retainAll(b);
        int[] res = new int[set.size()];
        int index = 0;
        for (int i : set) {
            res[index++] = i;
        }
        Arrays.sort(res);
        return res;
    }

    private List<Integer> toList(int[] array) {
        List<Integer> res = new ArrayList<>();
        for (int num1 : array) {
            res.add(num1);
        }
        return res;
    }
    // Time Complexity: O(m + n) in average, in worst case O(m * n)
    // Space Complexity: O(m + n)

    // Solution 2: two pointer, when the arrays are sorted
    public int[] intersection2(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        List<Integer> res = new ArrayList<>();
        int one = 0, two = 0;
        while (one < nums1.length && two < nums2.length) {
            int a = nums1[one], b = nums2[two];
            if (a < b) {
                one++;
            } else if (a > b) {
                two++;
            } else {
                res.add(a);
                // when have duplicated element
                // careful we should always remember to check the bound!!
                while (one < nums1.length && a == nums1[one]) {
                    one++;
                }
                while (two < nums2.length && b == nums2[two]) {
                    two++;
                }
            }
        }
        return listToArray(res);
    }

    private int[] listToArray(List<Integer> array) {
        int[] res = new int[array.size()];
        int index = 0;
        for (int i : array) {
            res[index++] = i;
        }
        return res;
    }

    // Time Complexity: O(m + n)
    // Space Complexity: O(1)

}